From: Andreas Gruenbacher <agruen@suse.de>
Subject: [FIX] Missed wakeup in pipe_readv
References: 50277

Chris Mason <mason@suse.com> has identified that the following changeset
introduced a bug that ultimately leads to a missed pthread cleanup
function call in glibc's tst-cancel17.c:

http://linux.bkbits.net:8080/linux-2.6/cset%401.1966.12.1?nav=index.html|ChangeSet@-8w

The fix is to move setting of the do_wakeup flag out of the if that
advances to the next buffer again.  Bug introduced in 2.6.10-bk11, and
tested against 2.6.10-bk11 and .

Signed-off-by: Andreas Gruenbacher <agruen@suse.de>

Index: linux-2.6.10-bk11-debug/fs/pipe.c
===================================================================
--- linux-2.6.10-bk11-debug.orig/fs/pipe.c
+++ linux-2.6.10-bk11-debug/fs/pipe.c
@@ -132,9 +132,9 @@ pipe_readv(struct file *filp, const stru
 				curbuf = (curbuf + 1) & (PIPE_BUFFERS-1);
 				info->curbuf = curbuf;
 				info->nrbufs = --bufs;
-				do_wakeup = 1;
 			}
 			total_len -= chars;
+			do_wakeup = 1;
 			if (!total_len)
 				break;	/* common path: read succeeded */
 		}
